gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSubImpl.java

```diff
@@ -445,17 +466,35 @@ public SubscriptionId apply(String compositeSubscription) {
 
   @Override
   public Iterator<ReceivedMessage> pull(String subscription, int maxMessages) {
-    // this should set return_immediately to true
-    return null;
+    return get(pullAsync(subscription, maxMessages));
   }
 
   @Override
-  public Future<Iterator<ReceivedMessage>> pullAsync(String subscription, int maxMessages) {
-    // though this method can set return_immediately to false (as future can be canceled) I
-    // suggest to keep it false so sync could delegate to asyc and use the same options
-    // this method also should use the VTKIT thread-pool to renew ack deadline for non consumed
-    // messages
-    return null;
+  public Future<Iterator<ReceivedMessage>> pullAsync(final String subscription, int maxMessages) {
+    PullRequest request = PullRequest.newBuilder().setReturnImmediately(true)
+        .setSubscription(SubscriberApi.formatSubscriptionName(options().projectId(), subscription))
+        .setMaxMessages(maxMessages)
+        .setReturnImmediately(true)
+        .build();
+    Future<PullResponse> response = rpc.pull(request);
```

Comment:
I don't think this is good enough. The whole notion of this async is get it in parallel and deal with it at some point. With this implementation that "some point" must be "almost immediately" otherwise all the messages might deadline might expire.

I think there are more ways to fix it. One way is to make the SPI layer expose the `ListenableFuture` as returned by the underlying VTKIT library or add our own version
of it (or use a Callback instead for this method).